euicc/hexutil.c

```diff
@@ -1,142 +1,110 @@
 #include "hexutil.h"
 
-#include <inttypes.h>
 #include <stdlib.h>
 #include <string.h>
 
-int euicc_hexutil_bin2hex(char *output, uint32_t output_len, const uint8_t *bin, uint32_t bin_len) {
-    const char hexDigits[] = "0123456789abcdef";
+#ifndef LIBEUICC_REDUCED_STDLIB_CALL
+#    include <ctype.h>
+#    include <stddef.h>
+#endif
 
-    if (!bin || !output) {
+inline int euicc_hexutil_hex2bin(uint8_t *output, const uint32_t output_len, const char *input) {
+    return euicc_hexutil_hex2bin_r(output, output_len, input, strlen(input));
+}
+
+int euicc_hexutil_hex2bin_r(uint8_t *output, const uint32_t output_len, const char *input, const uint32_t input_len) {
+    if (output == NULL || input == NULL || (input_len % 2) != 0 || output_len < (input_len / 2)) {
         return -1;
     }
+    uint32_t bytes = 0;
+#ifdef LIBEUICC_REDUCED_STDLIB_CALL
+    uint8_t c, msb = 0;
+    for (uint32_t i = 0; i < input_len; i++) {
+        c = input[i] | ' ';
+        if ((c < '0' || c > '9') && (c < 'a' || c > 'f')) {
+            return -1; // invalid character
+        }
+        c -= c > '9' ? 'a' - 10 : '0';
+        if (i % 2 == 0) {
+            msb = c;
+        } else {
+            output[bytes++] = msb << 4 | c;
+        }
+    }
+#else
+    char hex[3] = "\0\0";
+    for (uint32_t i = 0; i < input_len; i += 2) {
+        hex[0] = input[i + 0];
+        hex[1] = input[i + 1];
+        if (!(isxdigit(hex[0]) && isxdigit(hex[1]))) {
+            return -1; // invalid character
+        }
+        output[bytes++] = (uint8_t)strtol(hex, NULL, 16);
+    }
+#endif
+    return (int)bytes;
+}
 
-    if (output_len < 2 * bin_len + 1) {
+int euicc_hexutil_bin2hex(char *output, const uint32_t output_len, const uint8_t *bin, const uint32_t bin_len) {
+    if (output == NULL || bin == NULL || output_len < ((2 * bin_len) + 1)) {
         return -1;
     }
-
+    uint8_t msb, lsb;
+    uint32_t bytes = 0;
     for (uint32_t i = 0; i < bin_len; ++i) {
-        char byte = bin[i];
-        output[2 * i] = hexDigits[(byte >> 4) & 0x0F];
-        output[2 * i + 1] = hexDigits[byte & 0x0F];
+        msb = bin[i] >> 4;
+        lsb = bin[i] & 0xf;
+        output[bytes++] = (char)(msb + (msb > 9 ? 'a' - 10 : '0'));
+        output[bytes++] = (char)(lsb + (lsb > 9 ? 'a' - 10 : '0'));
```

Comment:
So many arithmetic operations, so many branch codes, it will be slower than the old LUT code, I think.